New component: zcash_address crate

str4d opened the pull request:
The idea is that `zcash_address` encapsulates parsing and serialization logic for Zcash address encodings, in such a way that downstream users:

- get automatic parsing support for future Zcash address formats simply by upgrading the dependency;
- get separate error messages for "invalid / unknown" vs "unsupported" addresses.
  - From past experience, these error messages are often just directly displayed to the user, so this helps to inform the user about why the address they pasted is not working.

Part of #371.

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,314 @@
+                        //   length 43, but if we added the simple form of detection keys
+                        //   the data would have length 75.
+                        return Err(
+                            if hrp.starts_with('z') && (data.len() == 43 || data.len() == 75) {
```
```suggestion
                            if hrp.starts_with('z') && (data.len() == 43 || data.len() == 64 || data.len() == 75) {
```

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,314 @@
+                        // - Zcash HRPs always start with a 'z'.
+                        // - Zcash shielded addresses with diversification have data of
+                        //   length 43, but if we added the simple form of detection keys
+                        //   the data would have length 75.
```
```suggestion
                        //   the data would have length 75. Alternatively if we switch from a
                        //   11-byte diversifier to two field elements, that would be 64 bytes.
```

daira commented on components/zcash_address/README.md:
```diff
@@ -0,0 +1,20 @@
+# zcash_address
+
+TBD
```
Zcash address parsing and serialization. This library allows its users to easily
recognize and give good error messages for new Zcash address types.

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,318 @@
+                    _ => {
+                        // Use some heuristics to try and guess whether this might be a Zcash
+                        // address from the future:
+                        // - Zcash HRPs always start with a 'z'.
```
It's quite possible that we may want to add transparent address formats that use bech32 encoding. I would suggest that we allow "t" followed by any single character as an HRP for transparent addresses.

daira reviewed (approved):
utACK.

Is it intended that this be able to parse address bundles/unified addresses? I think the API is already compatible with that.

str4d commented:
Yes; if/when we decide to use Unified Addresses, the Orchard key type would be replaced by UA parsing and serialization.

str4d commented:
- Rebased PR.
- Removed draft Orchard address encoding (which no longer exists).
- Added initial Unified Address encoding.
- Removed the future-address heuristics from this PR, as we now have the Unified Address format that we plan to use going forward.

I'm not sure whether this is the ideal API now that it includes UA support; I'm going to do some more thinking about it. But this should be mergeable as-is.

nuttycom commented on components/zcash_address/src/kind/unified.rs:
```diff
@@ -0,0 +1,141 @@
+                [typecode, length, data @ ..] if data.len() >= *length as usize => {
+                    let (addr, rest) = data.split_at(*length as usize);
+                    *encoded = rest;
+                    Some((*typecode, addr).try_into())
```
This `try_into` targets the `Receiver` type? This is one of those locations where type inference makes the code a bit less clear to read.

str4d replied:
Yes, this targets `Receiver` because the iterator is being collected into a `Vec<Receiver>` (being the type that `Address` wraps).

daira replied:
```suggestion
                    let receiver: Result<Receiver, _> = (*typecode, addr).try_into();
                    Some(receiver) 
```

nuttycom commented on components/zcash_address/src/kind/unified.rs:
```diff
@@ -0,0 +1,141 @@
+                    .chain(iter::repeat(0).take(PADDING_LEN))
+                    .collect();
+
+                f4jumble::f4jumble(&encoded).unwrap()
```
Blocking: I thought f4jumble was intended to be applied to the concatenated bytes (the entire unified address), not to each receiver?

str4d replied:
You are correct. The parser is also correct, so currently this isn't round-trip compatible. It happens to be passing tests though because we only have a test for a single-receiver address, for which this happens to work.

daira replied:
Remember that the input to F4Jumble also includes the 16 zero bytes.

str4d replied:
Yep, that was also a bug that I fixed at the same time.

nuttycom reviewed (changes requested):
I think that the point at which f4jumble is being applied in the encoding process is incorrect. Otherwise, this looks good.

nuttycom commented on components/zcash_address/src/convert.rs:
```diff
@@ -0,0 +1,175 @@
+    fn from_sprout(net: Network, data: sprout::Data) -> Self {
+        ZcashAddress {
+            net: if let Network::Regtest = net {
+                Network::Test
```
Addendum: I later found the comment where `Network::Regtest` was defined; I think that the correct thing to do is handle the network values separately at the use site, not to rewrite them in this location.

daira commented on components/zcash_address/LICENSE-MIT:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2021 Electric Coin Company
```
Electric Coin Company, or "Zcash developers"?

str4d replied:
ECC (matches the license text in `zcash_primitives/LICENSE-MIT` which is where the F4Jumble impl was moved from, and also matches ECC having copyright assignment for the rest of the code I've written in this module).

daira replied:
OK, but it'll need to be changed as soon as we get any external contributions, unless we intend to require copyright assignment (which we *don't* do for any of our other MIT/Apache code).

str4d replied:
We've been accepting external contributions for quite some time with the rest of the repo licensed in this way. If there's a change that needs making, it should be made uniformly.

daira replied:
I think this is violating the contributor's copyright, strictly speaking. Filed https://github.com/zcash/librustzcash/issues/393 .

nuttycom commented on components/zcash_address/src/kind/unified.rs:
```diff
@@ -120,22 +120,77 @@ impl TryFrom<&[u8]> for Address {
+    use super::{Address, Receiver};
+
+    prop_compose! {
+        fn uniform43()(a in uniform11(0u8..), b in uniform32(0u8..)) -> [u8; 43] {
```
Oh, this is a nice technique for larger byte arrays, I should have thought of this before! I've used `vec(any::<u8>(), size)` before but then had to convert.

str4d replied:
One downside to this is the separate halves shrink separately (though that's less of an issue here given that these addresses _are_ composite types). In theory we can implement our own custom array sizes directly, but I couldn't figure out how to set up the `UniformArrayStrategy` type properly.

nuttycom reviewed (approved):
utACK ff94f66d8e2b6e6efd89bfd2388d5a2c4c5887f2

daira commented on components/zcash_address/src/convert.rs:
````diff
@@ -0,0 +1,175 @@
+/// );
+/// ```
+pub trait ToAddress: private::Sealed {
+    fn from_sprout(net: Network, data: sprout::Data) -> Self;
+
+    fn from_sapling(net: Network, data: sapling::Data) -> Self;
+
````
New code is ideally supposed to be constructing only unified, P2SH, and P2PKH addresses, right?

str4d replied:
True; I added these for maximum compatibility (so you could replace your current parsing and serialization stack with `zcash_address`), but I guess until the old addresses go away, devs can use their existing serialization code.

daira replied:
I don't object to including them, but let's document that they are not preferred.

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+        // Remove leading and trailing whitespace, to handle copy-paste errors.
+        let s = s.trim();
+
+        // Most Zcash addresses use Bech32, so try that first.
```
Orchard uses bech32m, not bech32.

daira replied:
```suggestion
        // Most Zcash addresses use Bech32 or Bech32m, so try those first.
```

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+                    unified::MAINNET => Network::Main,
+                    unified::TESTNET => Network::Test,
+                    unified::REGTEST => Network::Regtest,
+                    // We will not define new Bech32m address encodings.
```
I think it's possible that we will. For instance, we may find a way to reduce the overhead currently imposed by the 16 bytes of redundancy, or some other way to mitigate address replacement attacks that doesn't use F4Jumble. Or there could be some cryptographic flaw in that design.

str4d replied:
Mmk, I'll remove this comment. But since there's no way to predict how we might make that kind of change, I can't do what I was previously doing (returning `ParseError::MaybeZcash`).

daira replied:
I think that we should reserve another Bech32m HRP for such purposes (just one should be sufficient).

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+}
+
+fn encode_b58(prefix: [u8; 2], data: &[u8]) -> String {
+    let mut decoded = Vec::with_capacity(2 + data.len());
```
Why is this called `decoded`? It's the byte sequence to be encoded.

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let encoded = match &self.kind {
+            AddressKind::Sprout(data) => encode_b58(
+                if let Network::Main = self.net {
+                    sprout::MAINNET
+                } else {
+                    sprout::TESTNET
```
```suggestion
                match self.net {
                    Network::Main => sprout::MAINNET,
                    Network::Test | Network::Regtest => sprout::TESTNET,
```

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+                &data.to_bytes(),
+            ),
+            AddressKind::P2pkh(data) => encode_b58(
+                if let Network::Main = self.net {
+                    p2pkh::MAINNET
+                } else {
+                    p2pkh::TESTNET
```
```suggestion
                match self.net {
                    Network::Main => p2pkh::MAINNET,
                    Network::Test | Network::Regtest => p2pkh::TESTNET,
```

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+                data,
+            ),
+            AddressKind::P2sh(data) => encode_b58(
+                if let Network::Main = self.net {
+                    p2sh::MAINNET
+                } else {
+                    p2sh::TESTNET
```
```suggestion
                match self.net {
                    Network::Main => p2sh::MAINNET,
                    Network::Test | Network::Regtest => p2sh::TESTNET,
```

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+                net: Network::Main,
+                kind: AddressKind::Unified(unified::Address(vec![unified::Receiver::Sapling(
+                    [0; 43],
+                )])),
```
Something is odd about the formatting of this being different from lines 233 and 240; was rustfmt re-run?

daira replied:
It appears that `rustfmt` (either run directly or via `cargo fmt`, or even `cargo fmt --all`) is not touching this code at all. I've seen this before for test code, and I don't know why. Maybe https://github.com/rust-lang/rustfmt/issues/3253 ? I didn't think that applied to tests.

daira commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,298 @@
+            "u1cd8yzk5mdn4n9r8c24tp8j8e9ethw3rr7ker5zhew3kycyyxggdzfkcq5f9yf2jv8m5ar8krncsntlfpx3p4azvwrkp8z74t3vu4kqq2",
+            ZcashAddress {
+                net: Network::Main,
+                kind: AddressKind::Unified(unified::Address(vec![unified::Receiver::Sapling(
+                    [0; 43],
+                )])),
```
```suggestion
                kind: AddressKind::Unified(unified::Address(vec![unified::Receiver::Sapling([0; 43])])),
```

daira commented on components/zcash_address/src/kind/unified.rs:
```diff
@@ -0,0 +1,196 @@
+            .chain(iter::repeat(0).take(PADDING_LEN))
+            .collect();
+
+        f4jumble::f4jumble(&encoded).unwrap()
```
I believe this could panic if the length of `encoded` were greater than 16448 bytes.

str4d replied:
Currently there is no way to construct a `unified::Address` other than decoding one (which checks the length), so this cannot panic. When we add other constructors, we can similarly enforce correctness there.

daira reviewed (approved):
utACK modulo previous comments.

str4d commented on components/zcash_address/src/encoding.rs:
```diff
@@ -0,0 +1,295 @@
+}
+
+fn encode_b58(prefix: [u8; 2], data: &[u8]) -> String {
+    let mut decoded = Vec::with_capacity(2 + data.len());
+    decoded.extend_from_slice(&prefix);
+    decoded.extend_from_slice(data);
+    bs58::encode(decoded).with_check().into_string()
```
```suggestion
    let mut bytes = Vec::with_capacity(2 + data.len());
    bytes.extend_from_slice(&prefix);
    bytes.extend_from_slice(data);
    bs58::encode(bytes).with_check().into_string()
```
